GH-574: Overall Connection Status of a Node with the MASQ Network

utkarshg6 opened the pull request.

utkarshg6 committed:
GH-574: feat - create initial skeleton for new node's connection

utkarshg6 committed:
GH-574: feat - add empty connect_to_masq_network()

utkarshg6 committed:
GH-574: feat - add overall_connection_status and it's impl block with tests

utkarshg6 committed:
GH-574: feat - add test and code for the remove()

utkarshg6 committed:
GH-574: refactor - remove unnecessary comments and update explanation for fields and variants

utkarshg6 committed:
GH-574: feat - add ConnectionProgressMessage to neighborhood.rs along with it's subscriber and handler

utkarshg6 committed:
GH-574: - add update_connection_stage() to overall_connection_status

utkarshg6 committed:
GH-574: trying to write test for connection_progress_message - switching to Bert

Commit:
GH-574: test established, implementation can begin

utkarshg6 committed:
GH-574: migrate AssertionMessage to test_utils/mod.rs and write code inside ConnectionProgressMessage handler

utkarshg6 committed:
GH-574: add ConnectionProgressMessage for TcpConnectionFailed

utkarshg6 committed:
GH-574: feat - test drive and change unwrap() to expect(); rename 'Internal Error' to 'Neighborhood Unbound'

utkarshg6 committed:
GH-574: rename tell_neighorhood and other small fixes

utkarshg6 committed:
GH-574: add AskAboutDebutGossipResponseMessage to neighborhood.rs

utkarshg6 committed:
GH-574: add the ability to panic in case we're skipping updation of stage

utkarshg6 committed:
GH-574: add boilerplate for handling AskAboutDebutGossipResponseMessage

utkarshg6 committed:
GH-574: test drive the event TcpConnectionFailed

utkarshg6 committed:
GH-574: improve test node_gossips_to_neighbors_on_startup

utkarshg6 committed:
feat: change NodeDescriptor to IpAddr for the overall_connection_status

utkarshg6 committed:
GH-574: add an expect in send_debut_gossip and eliminate a test

utkarshg6 committed:
GH-574: fix the expect statement by adding as_ref()

utkarshg6 committed:
GH-574: add ability to receive cpm_recipient inside GossipHandler via GossipAcceptorReal from Neighborhood; modify PassHandler to send CPM for Pass Gossip

utkarshg6 committed:
GH-574: add a test to verify whether bind message is populating gossip acceptor and producer; fixed another test

utkarshg6 committed:
GH-574: test drive the constructor for pass handler

utkarshg6 committed:
GH-574: remove previous_pass_targets from the overall_connection_status

utkarshg6 committed:
GH-574: Add Support for timestamp while handling pass gossip

utkarshg6 committed:
GH-574: migrate previous_pass_targets to gossip_acceptors and add tests for various cases

utkarshg6 committed:
GH-574: timestamp the pass targets received and update hashmap accordingly

utkarshg6 committed:
GH-574: allow neighborhood to handle CPM for Pass Gossips and OCS can update Connection Progress

utkarshg6 committed:
GH-574: neighborhood and OCS can handle CPM for event DeadEndFound

utkarshg6 committed:
GH-574: add the handler for AskAboutDebutGossipMessage; test event NoGossipResponseReceived

utkarshg6 committed:
GH-574: test drove CPM for Introduction Gossip

utkarshg6 committed:
GH-574: add a fn to check full neighborship by gossip source's ip

utkarshg6 committed:
GH-574: segregate standard gossips into 4 cases by checking full neighborship before and after db changes

utkarshg6 committed:
GH-574: test drive the cases for StandardGossipReceived, IntroductionGossipReceived and NoGossipResponseReceived

utkarshg6 committed:
GH-574: refactor the handle() for StandardGossipHandler

utkarshg6 committed:
GH-574: fix the failing test inside gossip_acceptor.rs

utkarshg6 committed:
GH-574: add the ability to update the stages of overall_connection_status

utkarshg6 committed:
GH-574: refactored the handle() of CPM

utkarshg6 committed:
GH-574: rename the fn from get_connection_progress() to get_connection_progress_by_ip()

utkarshg6 committed:
GH-574: refactor the handle_pass_gossip()

utkarshg6 committed:
GH-574: refactor tests in overall_connection_status.rs

utkarshg6 committed:
GH-574: add Node to UI Message named UiConnectionChangeBroadcast

utkarshg6 committed:
GH-574: send message to the ui for the advancing stages of overall connection

utkarshg6 committed:
GH-574: fix failing tests in neighborhood/mod.rs by providing recipient of node_to_ui and remove some compiler warnings

utkarshg6 committed:
GH-574: remove most of the compiler warnings

utkarshg6 committed:
GH-574: remove all the compiler warnings

utkarshg6 committed:
GH-574: add suffix _opt to variabled storing Option<T> in src/neighborhood/mod.rs

utkarshg6 committed:
GH-574: fix the clippy warnings

utkarshg6 committed:
GH-574: merge master

utkarshg6 committed:
GH-574: fix the test setup_results_are_broadcast_to_all_uis_integration

utkarshg6 committed:
GH-574: minor refactoring changes

utkarshg6 committed:
GH-574: fix ubuntu build by making the constant  public

utkarshg6 committed:
GH-574: change the panic message inside masq/tests/utils.rs

utkarshg6 committed:
GH-574: merge master

utkarshg6 committed:
GH-574: change interval for starting DaemonProcess in tests to 5s and remove relative thread sleeps

utkarshg6 committed:
GH-574: change the thread sleep to 10_000ms in the test provided_and_consumed_services_are_recorded_in_databases

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -560,14 +618,15 @@ impl GossipHandler for IntroductionHandler {
                 .expect("Introduction not properly qualified");
             let introducer_key = introducer.inner.public_key.clone();
-            match self.update_database(database, cryptde, introducer) {
+            match self.update_database(database, cryptde, introducer.clone()) {
```
There is a possibility not to have to clone it wholly. You could take just a copy of the ip address, e.g. creating it on a line before as a new variable, to use it in the connection progress message.

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -576,6 +635,15 @@ impl GossipHandler for IntroductionHandler {
                 }
             }
+            let connection_progess_message = ConnectionProgressMessage {
+                peer_addr: introducer.node_addr_opt.unwrap().ip_addr(),
```
You need to avoid using unwrap() in production. Use expect with a decent small message instead please.
(I usually use expectv() but that's just my vain of an author.)

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -576,6 +635,15 @@ impl GossipHandler for IntroductionHandler {
+            let connection_progess_message = ConnectionProgressMessage {
+                peer_addr: introducer.node_addr_opt.unwrap().ip_addr(),
+                event: ConnectionProgressEvent::IntroductionGossipReceived(
+                    introducee.node_addr_opt.as_ref().unwrap().ip_addr(),
```
Same here.

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -1991,6 +2214,165 @@ mod tests {
+
+    #[test]
+    fn no_cpm_is_sent_in_case_full_neighborship_doesn_t_exist_and_cannot_be_created() {
+        // (false, false)
```
Nobody will really know what this note was referring to. You can make it more obvious or just delete it.

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -1991,6 +2214,165 @@ mod tests {
+        let root_node = make_node_record(1111, true); // This is us
+        let mut root_db = db_from_node(&root_node);
+        // This node must not be neighbors with the root node.
+        let src_node = make_node_record(2222, true); // Full Neighbor
```
The comment is misleading. That's why not writing comments is usually better.

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -1991,6 +2214,165 @@ mod tests {
+        let cryptde = main_cryptde();
+        let root_node = make_node_record(1111, true); // This is us
+        let mut root_db = db_from_node(&root_node);
+        let src_node = make_node_record(2222, true); // Full Neighbor
```
I'd recommend to reconsider the use of comments like this one or if you want just remove it.

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -2540,27 +2922,218 @@ mod tests {
+    // 2. Send the AskDebutMessage
+    // 3. Waiting Period [Introduction Gossip, Pass Gossip, Standard Gossip] - Update stage on the basis of gossip
+    // 4. Receive AskDebutMessage {handle whether gossip was received or not.}
+    // 5. Update the stage to Failed in case not received
```
These notes may not be needed anymore.

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -2540,27 +2922,218 @@ mod tests {
     #[test]
     fn pass_is_properly_handled() {
+        // This test makes sure GossipAcceptor works correctly
+        // TODO: Make sure we test that PassHandler is called and not the others
```
Probably, what I do is that I deleted the second line with the TODO only. I think the concern has been gone by now.

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -2540,27 +2922,218 @@ mod tests {
+            ),
+        }
+        System::current().stop();
+        // System ran successfully
```
Please, this is too much of information :)

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -2540,27 +2922,218 @@ mod tests {
+        let pass_target_ip_addr = pass_target.node_addr_opt().unwrap().ip_addr();
+        subject.previous_pass_targets.borrow_mut().insert(
+            pass_target_ip_addr,
+            SystemTime::now().sub(PASS_GOSSIP_EXPIRED_TIME.add(Duration::from_secs(1))),
```
Maybe I'd recommend to create a variable, maybe even saying "excessive" or something to show clearly that the period of time is longer than PASS_GOSSIP_EXPIRED_TIME.

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -2774,7 +3348,7 @@ mod tests {
+        let subject = make_subject(main_cryptde());
         let original_dest_db = dest_db.clone();
 
         let result = subject.handle(
```
This is not your fault but could you change the order of items in the first assertion. "result" belongs to the left.

bertllll commented on node/src/neighborhood/gossip_acceptor.rs:
```diff
@@ -2679,6 +3252,7 @@ mod tests {
         dest_db.add_arbitrary_full_neighbor(dest_node.public_key(), third_node.public_key());
         src_db.add_arbitrary_full_neighbor(dest_node.public_key(), third_node.public_key());
         src_db.add_arbitrary_full_neighbor(src_node.public_key(), third_node.public_key());
+        src_db.add_arbitrary_full_neighbor(src_node.public_key(), dest_node.public_key());
```
I think this is wrong but cannot say why.
The sender probably shouldn't tell us we are a full neighbor if we had no knowledge a moment ago that the sender existed at all. We probably cannot jump from "no record" to "full neighbor" without tracking the node as a half neighbor before or sure, a full neighbor. :man_shrugging: 

Please reach the wise, Dan will help you.

bertllll commented on node/src/proxy_client/stream_handler_pool.rs:
```diff
@@ -1329,6 +1329,7 @@ mod tests {
 
     #[test]
     fn bad_dns_lookup_produces_log_and_sends_error_response() {
+        // TODO: This test fails sometimes for unknown reason, usually when a set of tests are run
```
You can spend some time on recreating the error here but I don't want to force you. Feel free to leave the comments there and we may fix it at some other time...or...well...never really.

utkarshg6 replied:
I wish we do it someday, I'm leaving it as it is for right now.

bertllll commented on node/src/stream_handler_pool.rs:
```diff
@@ -827,7 +858,7 @@ mod tests {
                 .unwrap();
 
-            sub_tx.send(subject_subs).expect("Internal Error");
+            sub_tx.send(subject_subs).expect("Neighborhood Unbound");
```
Please change it to unwrap() it's more convenient.

bertllll commented on node/src/sub_lib/neighborhood.rs:
```diff
@@ -499,10 +526,27 @@ impl fmt::Display for GossipFailure_0v1 {
+pub struct NeighborhoodTools {
+    pub notify_later_ask_about_gossip:
+        Box<dyn NotifyLaterHandle<AskAboutDebutGossipMessage, Neighborhood>>,
+    // TODO: Should we change the above field to constant
```
The comment actually should point to the field below. And it makes sense, you can create a constant meaning how long the interval for "ask about gossip" is. Then you can test the new constant in the test of all constants together.

bertllll commented on node/src/sub_lib/neighborhood.rs:
```diff
@@ -499,10 +526,27 @@ impl fmt::Display for GossipFailure_0v1 {
+    fn default() -> Self {
+        Self {
+            notify_later_ask_about_gossip: Box::new(NotifyLaterHandleReal::new()),
+            ask_about_gossip_interval: Duration::from_secs(10),
```
Here is the place where you'd put the constant. ;)

bertllll commented on node/src/sub_lib/neighborhood.rs:
```diff
@@ -531,6 +575,10 @@ mod tests {
             }
         );
+        assert_eq!(
+            NeighborhoodTools::default().ask_about_gossip_interval,
```
Here you will be testing the constant itself instead of a field of a struct.

bertllll commented on node/src/test_utils/neighborhood_test_utils.rs:
```diff
@@ -274,3 +279,19 @@ impl From<&NodeRecord> for AccessibleGossipRecord {
+    let recipient = addr.recipient();
+
+    (recipient, recording_arc)
+}
```
You could write:

```
fn make_recipient_and_recording<M:Message>()->(Recipient<M>,Arc<Mutex<Recording>)
{
    let (recorder, _, recording_arc) = make_recorder();
    let addr = recorder.start();
    let recipient = addr.recipient();

    (recipient, recording_arc)
}
```
Then you could stick it in two tiny functions like:
```
pub fn make_cpm_recipient() -> (Recipient<ConnectionProgressMessage>, Arc<Mutex<Recording>>) {
   make_recipient_and_recording()
}
```
...and the other, similarly...

utkarshg6 replied:
This is what the Rust compiler is accepting in the end:

```rust
pub fn make_recipient_and_recording_arc<M: 'static>() -> (Recipient<M>, Arc<Mutex<Recording>>)
where
    M: Message + std::marker::Send,
    <M as actix::Message>::Result: std::marker::Send,
    recorder::Recorder: actix::Handler<M>,
{
    let (recorder, _, recording_arc) = make_recorder();
    let addr = recorder.start();
    let recipient = addr.recipient::<M>();

    (recipient, recording_arc)
}

pub fn make_cpm_recipient() -> (Recipient<ConnectionProgressMessage>, Arc<Mutex<Recording>>) {
    make_recipient_and_recording_arc()
}

pub fn make_node_to_ui_recipient() -> (Recipient<NodeToUiMessage>, Arc<Mutex<Recording>>) {
    make_recipient_and_recording_arc()
}
```

Cons:
- The bounds are complex
- There is a `'static` lifetime

bertllll replied:
In general I don't mind to have those boundaries specified if they are needed. You can make the boundaries briefer like this:
 M: Message + Send,
 <M as actix::Message>::Result: Send,
'static

To be honest I don't understand this kind of boundary so please inspect if it is necessary...by temporarily removing it.    
Recorder: actix::Handler<M>,

bertllll reviewed (commented):
Remaining files to be reviewed:
node/src/neighborhood/mod.rs
node/src/neighborhood/overall_connection_status.rs

dnwiebe commented on node/src/neighborhood/mod.rs:
```diff
@@ -1565,14 +1626,500 @@ mod tests {
+        subject.tools.ask_about_gossip_interval = Duration::from_millis(10);
+        let addr = subject.start();
+        let cpm_recipient = addr.clone().recipient();
+        let beginning_connection_progress = ConnectionProgress {
```
There seems to be a lot of stuff here in the Arrange section of the test that's really about asserting, and it might be clearer if it were moved into the Assert part of the test.

dnwiebe commented on node/src/neighborhood/mod.rs:
```diff
@@ -1565,14 +1626,500 @@ mod tests {
+            prev_connection_progress: beginning_connection_progress.clone(),
+        };
+        let system = System::new("testing");
+        let assertions = Box::new(move |actor: &mut Neighborhood| {
```
This is strictly assertion-related; it seems that it should be in the Assert part of the test rather than the Arrange part.

dnwiebe commented on node/src/neighborhood/mod.rs:
```diff
@@ -1565,14 +1626,500 @@ mod tests {
+        let addr = subject.start();
+        let cpm_recipient = addr.clone().recipient();
+        let system = System::new("testing");
+        let assertions = Box::new(move |actor: &mut Neighborhood| {
```
It looks like this can also be moved down into the Assert section.

dnwiebe commented on node/src/neighborhood/mod.rs:
```diff
@@ -1565,14 +1626,500 @@ mod tests {
+                vec![ConnectionProgress {
+                    initial_node_descriptor: node_descriptor.clone(),
+                    current_peer_addr: node_ip_addr,
+                    connection_stage: ConnectionStage::Failed(DeadEndFound)
```
"DeadEndFound" might be clearer as "PassLoopFound"...or at least something that mentioned pass loops.

dnwiebe commented on node/src/neighborhood/mod.rs:
```diff
@@ -1565,14 +1626,500 @@ mod tests {
+        let addr = subject.start();
+        let cpm_recipient = addr.clone().recipient();
+        let system = System::new("testing");
+        // let new_pass_target = IpAddr::from_str("10.20.30.40").unwrap();
```
Commented-out code

dnwiebe commented on node/src/neighborhood/mod.rs:
```diff
@@ -2648,11 +3195,9 @@ mod tests {
-        let peer_actors = peer_actors_builder().build();
-        addr.try_send(BindMessage { peer_actors }).unwrap();
-        let sub = addr.recipient::<ExpiredCoresPackage<Gossip_0v1>>();
+        let sub = addr.clone().recipient::<ExpiredCoresPackage<Gossip_0v1>>();
```
Is this clone() necessary?

dnwiebe commented on node/src/neighborhood/mod.rs:
```diff
@@ -2847,10 +3392,13 @@ mod tests {
         let mut subject: Neighborhood = neighborhood_from_nodes(&subject_node, Some(&neighbor));
         let replacement_database = subject.neighborhood_database.clone();
-        subject.gossip_acceptor = Box::new(DatabaseReplacementGossipAcceptor {
+        // TODO: Make sure we modify this test and didn't write a different one testing the same thing
```
Leftover TODO

dnwiebe commented on node/src/neighborhood/mod.rs:
```diff
@@ -3533,15 +4082,16 @@ mod tests {
         addr.try_send(BindMessage { peer_actors }).unwrap();
-
-        let sub = addr.recipient::<StartMessage>();
+        let sub = addr.clone().recipient::<StartMessage>();
```
Unnecessary clone()?

dnwiebe commented on node/src/neighborhood/mod.rs:
```diff
@@ -4319,6 +4869,8 @@ mod tests {
         let persistent_config = PersistentConfigurationMock::new();
         subject.persistent_config_opt = Some(Box::new(persistent_config));
+        assert!(subject.gossip_acceptor_opt.is_none());
+        assert!(subject.gossip_producer_opt.is_none());
```
Probably one of these isn't necessary, unless there's some side-effect stuff going on in the background, in which case we need either comments (suboptimal) or redesign (preferred)!

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+            )
+        });
+        Self {
+            initial_node_descriptor: node_descriptor.clone(),
```
I don't think this clone is necessary.

utkarshg6 replied:
This one is required, because we can't move the value.

bertllll replied:
You could, I guess, create just an instance of IpAddr instead of the whole NodeAddr because one is Copy and the second is just Clone.

You can see that we later need only IpAddr so you can grab less than you do now, it's unnecessarily too a big piece of data.  

Changing...
` let node_addr = node_descriptor.node_addr_opt.as_ref().unwrap_or_else(|| {...`
 
 into...

` let ip_addr = node_descriptor.node_addr_opt.as_ref().unwrap_or_else(|| {blabla}).ip_addr();`

utkarshg6 replied:
Even after creating a new variable for `ip_addr`, it is not allowing me to move the `node_descriptor`. I guess, we'll have to keep a clone over here, I don't think we want to necessarily own the variable.

utkarshg6 replied:
This fn is now the owner of node_descriptor.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+                    "Can't update the stage from {:?} to {:?}",
+                    self.connection_stage, connection_stage
+                )
+            }
```
Please consider to refactor this place by removing the sequential check because it has mostly a value for the developer not for the software itself in order to work properly. It's kind of code which otherwise would be found in a test but here it's checked at runtime of the production code. It's usually a bad decision. 

That would also allow you to get rid of the TryFrom<usize> implementation which has just a very shallow purpose even now and then it would be useless absolutely.

utkarshg6 replied:
I can see the following two uses for this code:

1) It makes sure that transitions will only happen for the advancing stage.
2) It makes sure that we're not skipping any step, for example, going from `StageZero` to `NeighborshipEstablished` and skipping `TCPConnectionEstablished`.

According to me, it would be better if we keep it for now, and maybe add a TODO, saying

 "While working on regressing transitions, modify this function to allow both advancing and regressing transitions and also make sure no state is skipped during a transition"
 
 I would prefer if we open up this discussion, once the work on regressing transitions is completed, and we find these checks rudimentary at that point of time.

bertllll replied:
I showed you this as an option to avoid a runtime check by moving the procedure to tests and making it their responsibility to guarantee that we cannot take two steps instead of one. 

It would probably be some kind of enum:
    Up,
    Down,
    ZeroStage,
    Failure

Where I think you can test all these "orders", which would be supplied to your function, and the exact outcome of what the order caused. You could prove like that going by one step always when you call Up or Down. 
At special cases, though, you will need something more powerful and that's why there are those other options allowing longer movement. 

Also I would add that this solution would allow eliminating your specialized case of `.handle_pass_gossip()` which is structurally different from those respective cases in `.update_connection_stage()`. You could use `.update_stage` also there. 

Please make some considerably small TODO out of these notes so that we can get back to it when implementing the regression.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+enum OverallConnectionStage {
+    NotConnected = 0,        // Not connected to any neighbor.
+    ConnectedToNeighbor = 1, // Neighborship established. Same as No 3 hops route found.
+    ThreeHopsRouteFound = 2, // check_connectedness() returned true, data can now be relayed.
```
Please check that you have a single test for what numbers are linked to which variant of the enum. Then research whether those numbers couldn't be eliminated in the declaration because I think enums have a special property that they can automatically be converted into a number without explicitly defining the numbers. But only if you want a sequence like 0,1,2... of course.

utkarshg6 replied:
I searched for it, and every solution involves to explicitly enter the numbers.

Right now, we have a similar test, and I think since we only have three stages, this test is good enough.

```rust
    #[test]
    fn overall_connection_stage_can_be_converted_into_usize_and_can_be_compared() {
        assert!(
            OverallConnectionStage::ConnectedToNeighbor as usize
                > OverallConnectionStage::NotConnected as usize
        );
        assert!(
            OverallConnectionStage::ThreeHopsRouteFound as usize
                > OverallConnectionStage::ConnectedToNeighbor as usize
        );
    }
```

bertllll replied:
Let's do it in your favor. But you know it works the other way too :D

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+
+#[derive(PartialEq, Debug, Copy, Clone)]
+enum OverallConnectionStage {
+    NotConnected = 0,        // Not connected to any neighbor.
```
Consider removing the comments. The meaning is quite obvious at the enum.

utkarshg6 replied:
I have updated the comments, I believe the comments have potential to revive our memory in the late future, and will help us and a new developer to understand "How the implementation works?".


I don't have enough courage to remove these comments as I believe that the way code is written, everything may seem obvious to now, may not be in the future and definitely not to a new developer.

I truly think, these comments offer a helping hand, when someone is reading this piece of code.

bertllll replied:
Ok, I will leave you in peace. :))

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+    stage: OverallConnectionStage,
+    // Corresponds to progress with node descriptors entered by the user
+    pub progress: Vec<ConnectionProgress>,
+}
```
I think all three comments could easily go away. The programmer will figure it out from the context and names.

utkarshg6 replied:
I believe the same is true for these comments too.

bertllll replied:
I will look forward to the second review, you said it should be different then.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+            .find(|connection_progress| connection_progress.current_peer_addr == peer_addr)
+            .unwrap_or_else(|| {
+                panic!(
+                    "Unable to find the node in connections with IP Address: {}",
```
Maybe I wouldn't say node with a small case n. Or some other formulation?

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+            })
+            .unwrap_or_else(|| {
+                panic!(
+                    "Unable to find the node in connections with Node Descriptor: {:?}",
```
Maybe here too?

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+        node_to_ui_recipient: &Recipient<NodeToUiMessage>,
+    ) {
+        let connection_progress_to_modify = self.get_connection_progress_by_ip(peer_addr);
+
```
This is not mandatory but you could create a closure here to spare some space. 

let modify_connection_progress = | stage: ConnectionStage| connection_progress_to_modify.update_stage(stage); 

```
match event{
     ConnectionProgressEvent::TcpConnectionSuccessful  => modify_connection_progress(ConnectionStage::TcpConnectionEstablished),
```
     
You also can reimport ConnectionProgressEvent in your file like:
`use self::ConnectionProgressEvent::{TcpConnectionEstablished, TcpConnectionFailed,...etc.};     `

Then you can write the match statement as:

```
match event {
     TcpConnectionEstablished => <expression>,
```

It would also make the lines shorter and easier to read.

But again, it's not mandatory...you could try and decide by your liking

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+        let peer_1_ip = IpAddr::from_str("1.2.3.4").unwrap();
+        let peer_2_ip = IpAddr::from_str("5.6.7.8").unwrap();
+        let peer_3_ip = IpAddr::from_str("9.0.1.2").unwrap();
+
```
blank lines

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+        let desc_2 = make_node_descriptor_from_ip(peer_2_ip);
+        let desc_3 = make_node_descriptor_from_ip(peer_3_ip);
+
+        let initial_node_descriptors = vec![desc_1.clone(), desc_2.clone(), desc_3.clone()];
```
Consider to go with just two instances. It would look more compact.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+        let desc_1 = make_node_descriptor_from_ip(IpAddr::from_str("1.2.3.4").unwrap());
+        let desc_2 = make_node_descriptor_from_ip(IpAddr::from_str("5.6.7.8").unwrap());
+        let desc_3 = make_node_descriptor_from_ip(IpAddr::from_str("9.0.1.2").unwrap());
+
```
As above?

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+        let mut subject = OverallConnectionStatus::new(initial_node_descriptors);
+
+        let removed_desc_1 = subject.remove(0);
+        let removed_desc_2 = subject.remove(0);
```
Testing just one remove would suffice.

utkarshg6 replied:
I think it would be insufficient, we need to prove that the `remove()` works in the order First In, First Out when the `0` is passed as an argument.

bertllll replied:
I still think, you may disagree though, that just one line would be enough. With one difference, 
`let removed_desc_1 = subject.remove(0);` could be rather `let removed_desc = subject.remove(1);`, proving that the index matters. With your assertion, you'll show that you got the right descriptor back and the first can be ignored in the test regarding assertions.
BUT the test would be stronger if you also assert that subject.is_empty() is false to prove that .`remove(number)` did not erase everything.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+        let mut subject = OverallConnectionStatus::new(initial_node_descriptors);
+
+        subject.update_connection_stage(
+            node_descriptor.node_addr_opt.as_ref().unwrap().ip_addr(),
```
You can also use your already created variable node_ip_addr.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+        let mut subject = OverallConnectionStatus::new(initial_node_descriptors);
+
+        subject.update_connection_stage(
+            node_ip_addr.clone(),
```
This is actually Copy.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+                can_make_routes: false,
+                stage: OverallConnectionStage::NotConnected,
+                progress: vec![ConnectionProgress {
+                    initial_node_descriptor: node_descriptor.clone(),
```
The .clone() call shouldn't be necessary.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+                can_make_routes: false,
+                stage: OverallConnectionStage::ConnectedToNeighbor,
+                progress: vec![ConnectionProgress {
+                    initial_node_descriptor: node_descriptor.clone(),
```
Here too.

And then at couple of other tests. Just check them over.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+        let node_descriptor = make_node_descriptor_from_ip(node_ip_addr);
+        let initial_node_descriptors = vec![node_descriptor.clone()];
+        let mut subject = OverallConnectionStatus::new(initial_node_descriptors);
+        let new_node_ip_addr: IpAddr = Ipv4Addr::new(5, 6, 7, 8).into();
```
It may look nicer if you create the IpAddr the same way each time.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+    }
+
+    #[test]
+    fn updates_the_connection_stage_to_neighborship_established_when_standard_gossip_is_received() {
```
You may think of how to parametrize some of these tests (if you found just a small number of arguments needed to be supplied because of specifics of the given test), using common parts, reused.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+                }],
+            }
+        )
+    }
```
We probably lack a test proving that different tentacles cannot overwrite a more advanced stage achieved by the other one. I can imagine a threat with the reversed direction caused by answering a new pass gossip. That stimulates a step back to StageZero. Hopefully it's not possible but we also need a test making it obvious.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+    fn panics_at_updating_the_connection_stage_if_a_node_is_not_a_part_of_connections() {
+        let node_ip_addr = IpAddr::from_str("1.2.3.4").unwrap();
+        let node_descriptor = make_node_descriptor_from_ip(node_ip_addr);
+        let initial_node_descriptors = vec![node_descriptor.clone()];
```
This clone...

This shows up repeatedly at lines of the same purpose but in other tests.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+    }
+
+    #[test]
+    fn getter_fn_exists_for_boolean_can_make_routes() {
```
maybe like "we_can_ask_about_can_make_routes()"? Just a suggestion...avoiding the phrase "getter_fn..."

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+    }
+
+    #[test]
+    fn updates_connection_stage_to_failed_when_no_gossip_response_is_received() {
```
This part is frequently seen over here:

 ```
   let node_ip_addr = IpAddr::from_str("1.2.3.4").unwrap();
        let node_descriptor = make_node_descriptor_from_ip(node_ip_addr);
        let initial_node_descriptors = vec![node_descriptor.clone()];
        let mut subject = OverallConnectionStatus::new(initial_node_descriptors);
        let (recipient, _) = make_node_to_ui_recipient();
```
I believe you could put it inside an outboard method, a utility function.

utkarshg6 replied:
A new utility fn is added `make_node_and_recipient()`.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+    }
+
+    #[test]
+    fn updates_the_stage_to_three_hops_route_found_in_case_introduction_gossip_is_received_and_flag_is_true(
```
These two tests below can be easily parametrized. That's one point. But we think we'd rather see four tests, one for each of four possible states in the relation between: can_make_routes (bolean) X ConnectionProgressEvent::{IntroductionGossipReceived, StandardGossipReceived}.

bertllll commented on node/src/neighborhood/overall_connection_status.rs:
```diff
@@ -0,0 +1,960 @@
+    }
+
+    #[test]
+    fn doesn_t_send_message_to_the_ui_in_case_gossip_is_received_but_stage_hasn_t_updated() {
```
This one and the test below could be easily and heavily parametrized. Here I see it very straight. 

By a swift glimpse, this could take only two parameters, one is the stage in the subject, the second is the ConnectionProgressEvent passed into the act of the test.     

You may create variables for the two parameters to have a chance to give them names. The names could reveal what they stand for to the reader. Because a picture of a function with two inlined parameters without any further explanation is not so encouraging for me in the question of benefits of this solution.

utkarshg6 replied:
I don't think it's worth the struggle. Even though parameterising will make the size of these functions short, it would encapsulate the assertions, and will work for only two tests.

I agree that the differences are minor and it's possible to parameterise, but I believe that the trade off isn't worth it.

What I would recommend is to minimise the possible lines and not to go with the parameterisation.

utkarshg6 committed:
GH-574: add review changes for gossip_acceptor.rs

utkarshg6 committed:
GH-574: add review changes for neighborhood.rs and stream_handler_pool.rs

utkarshg6 committed:
GH-574: refactor tests in neighborhood/mod.rs

utkarshg6 committed:
GH-574: rename DeadEndFound to PassLoopFound

utkarshg6 committed:
GH-574: add review changes for src/neighborhood/mod.rs

utkarshg6 committed:
GH-574: add review changes for overall_connection_status.rs and neighborhood_test_utils.rs

utkarshg6 committed:
GH-574: add a generalized fn for building recipients

utkarshg6 committed:
GH-574: fix the handle() of StandardGossip

utkarshg6 committed:
GH-574: add test progress_done_by_one_connection_progress_can_not_be_overridden_by_other_in_overall_connection_progress

utkarshg6 committed:
GH-574: parameterize test regarding gossip and can_make_routes bool into 4 tests

utkarshg6 committed:
GH-574: make ConnectionProgress owner of initial_node_descriptors and remove clone

utkarshg6 committed:
GH-574: add some helper functions in overall_connection_status.rs

utkarshg6 committed:
GH-574: make helper function make_node_and_recipient()

utkarshg6 committed:
GH-574: migrate helper functions to neighborhood_test_utils.rs

utkarshg6 committed:
GH-574: refactored the helper fn result_when_neighborship_is_established_and_can_make_routes_is_updated

utkarshg6 committed:
GH-574: refactor two more tests that doesn't sends any message to the UI

utkarshg6 committed:
GH-574: use make_ip() to create a new ip for tests

utkarshg6 committed:
GH-574: add minor refactor changes to overall_connection_status.rs

utkarshg6 committed:
GH-574: rename utility fn to make_descriptor()

utkarshg6 committed:
GH-574: rename fn names and make pub key using string name in fn make_node_descriptor() (#142)

utkarshg6 committed:
Gh 574 merge master (#144)

* Docs/contributing plus cleanup (#138)

* update Practice with reference to version & QA

* update practice with version release info

* DRAFT: Update docs/branch (#133)

* Update node top level README

add list of component README files for easy navigation

* Added reference to UI-Interface-Interface

Added reference to UI-Interface-Interface document that lives in the Node repo main folder

* Update README.md

* Delete SHRD-faucet

This is no longer usable and is not connected to the MASQ token contracts

* Update to blockchain service details 

Update to blockchain service details about Polygon, testnets, Alchemy 
also removed old images

* wording and formattting

* update LICENSE

updated with corrected copyright for MASQ, and added info regarding previous codebase being removed from GitHub and original repo disappeared

* Update Practice with QA blurb

Thanks Stew

* adjust copyright

* remove old DNS doc

* remove old DNS doc

* remove old DNS doc

* delete old faucet info

* updated with component readmes

* remove old logo

* remove old logo2

* Create brand-guidelines.md

* add masq logos

* remove to move logos

* add masq logos

* added masq logo to readme

* update from reviews

* update and copyright bump

* copyright bump [skip ci]

[skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* docs/contributing-plus-cleanup: last polishment

Co-authored-by: KauriHero <55721017+kauri-hero@users.noreply.github.com>
Co-authored-by: Bert <65427484+bertllll@users.noreply.github.com>
Co-authored-by: Bert <Bert@Bert.com>

* GH-597: correct AutomapProtocol db storage (#120)

* GH-597: code and tests established; some new features or work style used: no_share_test wider use

* GH-597: little corrections and formatting

* GH-597: auto review

* GH-597: cosmetics, added a few places with logs

* GH-597: A rookie mistake :) not running and so forgetting integration tests

* update Practice with reference to version & QA

* update practice with version release info

* DRAFT: Update docs/branch (#133)

* Update node top level README

add list of component README files for easy navigation

* Added reference to UI-Interface-Interface

Added reference to UI-Interface-Interface document that lives in the Node repo main folder

* Update README.md

* Delete SHRD-faucet

This is no longer usable and is not connected to the MASQ token contracts

* Update to blockchain service details 

Update to blockchain service details about Polygon, testnets, Alchemy 
also removed old images

* wording and formattting

* update LICENSE

updated with corrected copyright for MASQ, and added info regarding previous codebase being removed from GitHub and original repo disappeared

* Update Practice with QA blurb

Thanks Stew

* adjust copyright

* remove old DNS doc

* remove old DNS doc

* remove old DNS doc

* delete old faucet info

* updated with component readmes

* remove old logo

* remove old logo2

* Create brand-guidelines.md

* add masq logos

* remove to move logos

* add masq logos

* added masq logo to readme

* GH-597: backup; closely before finishing / fixed an old issue with annoying 'Send' requirement; introducet a new concept of an arbitrary_id_stamp for testing specific situations

* GH-597: finished the concept of ArbitraryIdStamp

* GH-597: updated github yml files; the release one is now more optimized

* GH-597: last polishing before another review

* GH-597: second review addressed

* update from reviews

* update and copyright bump

* copyright bump [skip ci]

[skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* copyright bump [skip ci]

* docs/contributing-plus-cleanup: last polishment

* GH-597: a little correction in docs

* GH-597: fixed a labile test in WebsocketSuperwisor and added an explanation of an unclear solution around there too

* GH-597: removing unknown extern crate usage

* GH-597: trying to get the release build for Node working properly

* GH-597: removing an optimized build script because of an issue on Mac; we don't know how to fix that at the moment

Co-authored-by: Bert <Bert@Bert.com>
Co-authored-by: KauriHero <55721017+kauri-hero@users.noreply.github.com>

* GH-574: modify ci-matrix.yml

* GH-574: allow multinode integration tests to fail in actions

Co-authored-by: FinsaasGH <89403560+FinsaasGH@users.noreply.github.com>
Co-authored-by: KauriHero <55721017+kauri-hero@users.noreply.github.com>
Co-authored-by: Bert <65427484+bertllll@users.noreply.github.com>
Co-authored-by: Bert <Bert@Bert.com>

utkarshg6 committed:
Merge branch 'master' into GH-574

utkarshg6 committed:
GH-574: add todo for the fix of the failing test in GH Actions

utkarshg6 committed:
GH-574: change get_connection_progress_by_ip() to return a Result

utkarshg6 committed:
GH-574: migrate updation of OCS Stage from update_connection_stage() to handler for ConnectionProgressMessage

utkarshg6 committed:
GH-574: migrate the test to check whether two cpm's progress override each other to neighborhood/mod.rs

utkarshg6 committed:
GH-574: Improve the tests in Neighborhood

utkarshg6 committed:
GH-574: Bert invented a method in Recorder to stop the system after a certain type of message is received.

utkarshg6 committed:
GH-574: modify make_recipient_and_recording_arc() to accept stopping_message

utkarshg6 committed:
GH-574: use hashmap for the stop condition inside the Recorder

utkarshg6 committed:
GH-574: use stopping_message inside the tests of neighborhood

utkarshg6 committed:
GH-574: refactor tests in Overall Connection Status to test whether a message is sent or not and updates the stage

utkarshg6 committed:
GH-574: eliminate warnings, eprintlns and comments

utkarshg6 committed:
GH-574: eliminate the helper fn make_node_and_recipient()

utkarshg6 committed:
GH-574: add safety measure to recorder and refactor tests for handling CPM in Neighborhood

utkarshg6 committed:
GH-574: strengthen tests for handling CPM by asserting on the complete struct of OverallConnectionStatus

utkarshg6 committed:
GH-574: remove comment

utkarshg6 committed:
GH-574: improve test for tcp connection failed

utkarshg6 committed:
GH-574: activate the script for ci-matrix.yml

utkarshg6 committed:
GH-574: change chrono to time

utkarshg6 committed:
GH-574: fix the problem in receiving OffsetDateTime in log() of TestLogger

utkarshg6 committed:
GH-574: migrate the code related to recorder in a different module

utkarshg6 committed:
GH-574: use time instead of chrono in the test setup_results_are_broadcast_to_all_uis_integration

utkarshg6 committed:
GH-574: review changes

utkarshg6 committed:
GH-574: extend tests to assert whether neighborhood logs when it receives messages from unknown ip or desc

utkarshg6 committed:
GH-574: migrate the function real_format_function() to a different file

utkarshg6 committed:
GH-574: fix test transmit_log_handles_overloading_by_sending_msgs_from_multiple_threads

utkarshg6 committed:
Merge remote-tracking branch 'upstream/master' into GH-574

utkarshg6 committed:
GH-574: fix clippy warnings

utkarshg6 committed:
GH-574: add more rpc urls for polygon mainnet

utkarshg6 committed:
GH-574: ignore the test in logger.rs - transmit_log_handles_overloading_by_sending_msgs_from_multiple_threads

utkarshg6 committed:
GH-574: change the precision of time to milliseconds in logs

utkarshg6 committed:
GH-574: remove ignore attribute from the test transmit_log_handles_overloading_by_sending_msgs_from_multiple_threads

utkarshg6 committed:
GH-574: change duration from 1ms to 10ms in overloading_function

utkarshg6 committed:
GH-574: display messages on the terminal screen

dnwiebe committed:
Added multinode test

utkarshg6 committed:
GH-622: Add Connection Status Interrogation (Conversation Message) and introduce logs on connection stage update (#158)

* GH-622: introduce new command named connection-status and related request and response messages

* GH-622: sent stdout and stderr messages for the Connection Status Command

* GH-622: handle the connection status message inside neighborhood

* GH-622: write logs in debug mode when a debut gossip is sent or any connection stage gets updated

* GH-622: send connection status message to the specific client instead of all clients

* GH-622: add a multinode test connection_progress_can_be_requested

* GH-622: Very minor changes

* Removed comment

* GH 622 Review 1 (#160)

* GH-622: change the messages for connected to neighbor and three hops route found inside connection_status_command.rs

* GH-622: migrate the test connection_status_command_works() to command_factory.rs

* GH-622: write a helper function named get_transact_params_and_std_output_for_connection_status()

* GH-622: rename the todo inside connection_change_notification.rs

* GH-622: split the debut gossip logs to debug and trace.

* GH-622: Change the debug log message for connection stage update of pass gossip.

* GH-622: add docs for connectionStatus command to the USER-INTERFACE-INTERFACE.md

* Gh 622 Review 2 (#161)

* GH-622: fix typo inside the debug log of handle pass gossip

* GH-622: refactor the tests to make them inline in connection_status_command.rs

* GH-622: add clap subcommand code for connection status and introduce the command in schema.rs

* GH-622: remove clippy warnings

* GH-622: capitalize first letter in Node wherever it is used in logs

Co-authored-by: Dan Wiebe <dnwiebe@gmail.com>

utkarshg6 committed:
Merge branch 'master' into GH-574

utkarshg6 committed:
GH-574: remove warnings

dnwiebe committed:
GH-574: Trying to figure out why connection status broadcasts have stopped

dnwiebe committed:
Added consuming wallet to multinode test

utkarshg6 committed:
GH-574: add formatting

dnwiebe committed:
I think I have StreamHandlerPool right

dnwiebe committed:
Merge branch 'GH-574' of github.com:MASQ-Project/Node into GH-574

dnwiebe committed:
Formatting, mostly

dnwiebe committed:
Trying to get pnet compiling

dnwiebe committed:
Revert "Trying to get pnet compiling"

This reverts commit 385003d19ae1dbf044206935484e22a6d2d2284c.

dnwiebe committed:
Revert "Formatting, mostly"

This reverts commit ceb2a6f44f3bc912a343b9756fe8c1244cd6538c.

dnwiebe committed:
Revert "Merge branch 'GH-574' of github.com:MASQ-Project/Node into GH-574"

This reverts commit 467be879308f398662259c7a51859d71e396c4d8, reversing
changes made to 70a1fdf197ee8bfa63d25a702ad01b129e9f806d.

dnwiebe committed:
Revert "I think I have StreamHandlerPool right"

This reverts commit 70a1fdf197ee8bfa63d25a702ad01b129e9f806d.

dnwiebe committed:
Added three more Nodes in a multinode test

dnwiebe committed:
More imports

dnwiebe committed:
Added some logging

utkarshg6 committed:
GH-574: Update OCS stage and send the message to the UI from check_connectedness()

Commit:
GH-574: fixed the test util in multinode tests that waits for messages for the UI; Utkarsh's test finally passing :)

Commit:
GH-574: I forgot to include one more file to the previous commit

utkarshg6 committed:
GH-574: add debug log when the stage of OCS has changed and message is sent to the UI.

utkarshg6 committed:
GH-574: Refactor of the function update_ocs_stage_and_send_message_to_ui() and improved logging.

utkarshg6 committed:
GH-574: improve logging for check_connectedness() and write tests for it.

utkarshg6 committed:
GH-574: refactor the function that returns neighborhood linearly connected to nodes

utkarshg6 committed:
GH-574: add test for debug log when the check_connectedness() hasn't found a 3 hops route

utkarshg6 committed:
GH: add some formatting

utkarshg6 committed:
Merge branch 'master' into GH-574

utkarshg6 committed:
GH-574: remove clippy warnings exclusive to the current card

utkarshg6 committed:
GH-574: review 7 changes

utkarshg6 committed:
Merge branch 'master' into GH-574
